File: Sources/DiskMapCore/AppLeftoverFinder.swift
import Foundation

public struct AppLeftoverEntry: Sendable, Equatable {
    public let url: URL
    public let bytes: Int64

    public init(url: URL, bytes: Int64) {
        self.url = url
        self.bytes = bytes
    }
}

public struct AppLeftovers: Sendable {
    public let bundleID: String?
    public let appName: String
    public let bundlePath: URL
    public let bundleSize: Int64
    public let leftoverPaths: [URL]
    public let leftoverEntries: [AppLeftoverEntry]
    public let leftoverSize: Int64

    public init(
        bundleID: String?,
        appName: String,
        bundlePath: URL,
        bundleSize: Int64,
        leftoverPaths: [URL],
        leftoverEntries: [AppLeftoverEntry] = [],
        leftoverSize: Int64
    ) {
        self.bundleID = bundleID
        self.appName = appName
        self.bundlePath = bundlePath
        self.bundleSize = bundleSize
        self.leftoverPaths = leftoverPaths
        self.leftoverEntries = leftoverEntries.isEmpty
            ? leftoverPaths.map { AppLeftoverEntry(url: $0, bytes: 0) }
            : leftoverEntries
        self.leftoverSize = leftoverSize
    }
}

/// Finds files an app scattered outside its own .app bundle when it was
/// installed/run — caches, preferences, saved state, containers, logs.
///
/// There is no Apple API that enumerates "everything this app touched."
/// Every open-source uninstaller in this space (AppCleaner, Pearcleaner,
/// this one) does the same heuristic: search known Library locations for
/// items whose name contains the app's bundle identifier (reverse-DNS,
/// e.g. "com.example.app" — rarely collides by accident) or, failing
/// that, the app's display name (fuzzier, more false-positive risk).
///
/// Because it's heuristic, leftovers should always land in a *staged*
/// cleanup list for the user to review — never delete automatically.
public enum AppLeftoverFinder {

    static let searchLocations: [String] = [
        "~/Library/Caches",
        "~/Library/Application Support",
        "~/Library/Preferences",
        "~/Library/Saved Application State",
        "~/Library/Containers",
        "~/Library/Group Containers",
        "~/Library/HTTPStorages",
        "~/Library/WebKit",
        "~/Library/Logs",
        "~/Library/LaunchAgents",
        "/Library/LaunchDaemons",
        "/Library/Application Support",
    ]

    public static func defaultApplicationDirectories() -> [URL] {
        [
            URL(fileURLWithPath: "/Applications", isDirectory: true),
            URL(fileURLWithPath: NSHomeDirectory() + "/Applications", isDirectory: true),
        ]
    }

    public static func applicationBundles(in directories: [URL]) -> [URL] {
        let fm = FileManager.default
        var apps: [URL] = []
        for directory in directories {
            guard let contents = try? fm.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: nil,
                options: [.skipsHiddenFiles]
            ) else { continue }
            apps.append(contentsOf: contents.filter { $0.pathExtension == "app" })
        }
        return apps.sorted { $0.lastPathComponent.localizedStandardCompare($1.lastPathComponent) == .orderedAscending }
    }

    public static func findLeftovers(for appURL: URL, measureRelated: Bool = true) -> AppLeftovers {
        let bundle = Bundle(url: appURL)
        let bundleID = bundle?.bundleIdentifier
        let appName = appURL.deletingPathExtension().lastPathComponent

        var matches: [URL] = []
        let fm = FileManager.default

        for location in searchLocations {
            let expanded = (location as NSString).expandingTildeInPath
            let dirURL = URL(fileURLWithPath: expanded)
            guard let contents = try? fm.contentsOfDirectory(at: dirURL, includingPropertiesForKeys: nil) else {
                continue
            }

            for item in contents {
                let itemName = item.lastPathComponent

                if let bid = bundleID, itemName.localizedCaseInsensitiveContains(bid) {
                    matches.append(item)
                } else if bundleID == nil, appName.count > 3,
                          itemName.localizedCaseInsensitiveContains(appName) {
                    matches.append(item)
                }
            }
        }

        let bundleSize = allocatedSize(of: appURL)
        var entries: [AppLeftoverEntry] = []
        var leftoverSize: Int64 = 0
        if measureRelated {
            for url in matches {
                let bytes = allocatedSize(of: url)
                leftoverSize += bytes
                entries.append(AppLeftoverEntry(url: url, bytes: bytes))
            }
        } else {
            entries = matches.map { AppLeftoverEntry(url: $0, bytes: 0) }
        }

        return AppLeftovers(
            bundleID: bundleID,
            appName: appName,
            bundlePath: appURL,
            bundleSize: bundleSize,
            leftoverPaths: matches,
            leftoverEntries: entries,
            leftoverSize: leftoverSize
        )
    }

    public static func allocatedSize(of url: URL) -> Int64 {
        guard let enumerator = FileManager.default.enumerator(
            at: url, includingPropertiesForKeys: [.totalFileAllocatedSizeKey]
        ) else { return 0 }

        var total: Int64 = 0
        for case let fileURL as URL in enumerator {
            if let size = try? fileURL.resourceValues(forKeys: [.totalFileAllocatedSizeKey]).totalFileAllocatedSize {
                total += Int64(size)
            }
        }
        return total
    }
}

File: windows/src/DiskMap.Core/AppLeftoverFinder.cs
using DiskMap.Core.Native;
using Microsoft.Win32;

namespace DiskMap.Core;

public sealed record AppLeftovers(
    string? AppId,
    string AppName,
    string InstallPath,
    long InstallSize,
    List<string> LeftoverPaths,
    long LeftoverSize);

public sealed record InstalledApp(string Name, string? Publisher, string? InstallLocation, string? RegistryKeyName);

/// <summary>
/// Finds files an app scattered outside its install directory when it was
/// installed/run — caches, preferences, saved state, logs.
///
/// Windows has no API that enumerates "everything this app touched"
/// either. The heuristic mirrors the macOS build: enumerate installed apps
/// from the registry Uninstall hives (HKLM x64 + 32-bit view + HKCU), then
/// search the standard user/system data locations for items whose name
/// contains the app's display name or publisher token. Because it's
/// heuristic, leftovers always land in a *staged* cleanup list for review —
/// never delete automatically.
/// </summary>
public static class AppLeftoverFinder
{
    private const string UninstallPath = @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall";

    private static string[] SearchLocations =>
    [
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),      // %APPDATA%
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), // %LOCALAPPDATA%
        Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),// %PROGRAMDATA%
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Documents"),
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs"),
    ];

    /// <summary>
    /// Installed apps from the three Uninstall hives: HKLM 64-bit,
    /// HKLM 32-bit (Wow6432Node), HKCU. Same role as scanning
    /// /Applications on macOS.
    /// </summary>
    public static List<InstalledApp> InstalledApplications()
    {
        var apps = new List<InstalledApp>();
        ReadUninstallHive(RegistryHive.LocalMachine, RegistryView.Registry64, apps);
        ReadUninstallHive(RegistryHive.LocalMachine, RegistryView.Registry32, apps);
        ReadUninstallHive(RegistryHive.CurrentUser, RegistryView.Registry64, apps);
        return apps
            .GroupBy(a => a.Name, StringComparer.OrdinalIgnoreCase)
            .Select(g => g.First())
            .OrderBy(a => a.Name, StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    private static void ReadUninstallHive(RegistryHive root, RegistryView view, List<InstalledApp> apps)
    {
        using var baseKey = RegistryKey.OpenBaseKey(root, view);
        using var uninstall = baseKey.OpenSubKey(UninstallPath);
        if (uninstall is null) return;
        foreach (var subName in uninstall.GetSubKeyNames())
        {
            using var sub = uninstall.OpenSubKey(subName);
            if (sub is null) continue;
            var name = sub.GetValue("DisplayName") as string;
            if (string.IsNullOrWhiteSpace(name)) continue;
            // System components and updates aren't user-uninstallable apps.
            if (sub.GetValue("SystemComponent") is int sc && sc == 1) continue;
            apps.Add(new InstalledApp(
                name.Trim(),
                (sub.GetValue("Publisher") as string)?.Trim(),
                (sub.GetValue("InstallLocation") as string)?.Trim(),
                subName));
        }
    }

    public static AppLeftovers FindLeftovers(InstalledApp app)
    {
        var matches = new List<string>();
        // Match tokens: display name and publisher are the closest thing
        // Windows has to a bundle identifier.
        var tokens = new List<string> { app.Name };
        if (!string.IsNullOrWhiteSpace(app.Publisher) && app.Publisher.Length > 3)
            tokens.Add(app.Publisher);

        foreach (var location in SearchLocations)
        {
            if (!Directory.Exists(location)) continue;
            IEnumerable<string> contents;
            try { contents = Directory.EnumerateFileSystemEntries(location); }
            catch { continue; }

            foreach (var item in contents)
            {
                string itemName = Path.GetFileName(item);
                if (tokens.Any(t => t.Length > 3 && itemName.Contains(t, StringComparison.OrdinalIgnoreCase)))
                {
                    // Don't match the app's own install dir or the Recycle Bin.
                    if (item.Contains("$Recycle.Bin", StringComparison.OrdinalIgnoreCase)) continue;
                    matches.Add(item);
                }
            }
        }

        long installSize = !string.IsNullOrWhiteSpace(app.InstallLocation) && Directory.Exists(app.InstallLocation)
            ? AllocatedSize(app.InstallLocation)
            : 0;
        long leftoverSize = matches.Sum(AllocatedSize);

        return new AppLeftovers(
            AppId: app.RegistryKeyName,
            AppName: app.Name,
            InstallPath: app.InstallLocation ?? "",
            InstallSize: installSize,
            LeftoverPaths: matches,
            LeftoverSize: leftoverSize);
    }

    public static long AllocatedSize(string path)
    {
        try
        {
            var info = new FileInfo(path);
            if (!info.Exists && !Directory.Exists(path)) return 0;
            if ((info.Attributes & FileAttributes.Directory) == 0)
                return OnDiskSize(path, info.Length, info.Attributes);

            long total = 0;
            foreach (var entry in Directory.EnumerateFileSystemEntries(path, "*",
                         SearchOption.AllDirectories))
            {
                try
                {
                    var fi = new FileInfo(entry);
                    if ((fi.Attributes & FileAttributes.Directory) != 0) continue;
                    total += OnDiskSize(entry, fi.Length, fi.Attributes);
                }
                catch { /* skip unreadable */ }
            }
            return total;
        }
        catch { return 0; }
    }

    private static long OnDiskSize(string path, long logical, FileAttributes attributes)
    {
        const FileAttributes special = FileAttributes.Compressed | FileAttributes.SparseFile;
        if ((attributes & special) != 0)
        {
            uint low = Win32.GetCompressedFileSizeW(Win32.ExtendedPath(path), out uint high);
            if (low != 0xFFFFFFFF) return ((long)high << 32) | low;
        }
        if (logical == 0) return 0;
        // Round up to a 4K cluster — the common NTFS size; precise per-file
        // clusters would need per-volume queries that don't pay for a
        // heuristic leftover listing.
        return (logical + 4095) / 4096 * 4096;
    }
}
